server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class DesiredNodesMembershipService implements ClusterStateListener {
+    private final Set<DesiredNode> members;
+    private String latestHistoryId = null;
+
+    DesiredNodesMembershipService() {
+        this.members = new HashSet<>();
+    }
+
+    public static DesiredNodesMembershipService create(ClusterService clusterService) {
+        var tracker = new DesiredNodesMembershipService();
+        clusterService.addListener(tracker);
+        return tracker;
+    }
+
+    @Override
+    public synchronized void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
```

Comment:
I wonder if we should just maintain this always, regardless of who is master? That would ensure it is readily available on failover. Otherwise, I worry that we need another special case for the failover case, i.e., when node was not previously master, since we only consider the delta of nodes below.

Reply:
Yes, I think it should be fine if we keep track of this in all nodes and as you mention it should simplify things